internal/xpkg/layers.go

```diff
@@ -156,3 +159,106 @@ func AnnotateLayers(i v1.Image) (v1.Image, error) {
 
 	return mutate.ConfigFile(img, cfgFile)
 }
+
+// ImageFromFiles creates a v1.Image from arbitrary files on disk.
+// Each top-level directory at `root` is a separate layer.
+// The function performs no interpretation (parsing) of the files.
+func ImageFromFiles(root string) (v1.Image, error) {
+	extManifest := empty.Image
+
+	entries, err := os.ReadDir(root)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+
+		curDir := filepath.Join(root, entry.Name())
+
+		// Since there is an arbitrary directory of mostly small files, we'll
+		// forego streaming in-memory at the expense of some disk I/O with temporary tarballs.
+		tmpFile, err := os.CreateTemp("", "extension-*.tar")
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = tmpFile.Close() }()
```

Comment:
Nit: I'd prefer to avoid the anonymous function, given we're discarding the error anyway. 

```suggestion
		defer tmpFile.Close() //nolint:errcheck // Some explanation
```

That said, given that we're writing this file we could in theory hit a real error we might want to report to the user, right? e.g. If the file can't be flushed on close.

Reply:
Also, do we need to delete the temp tarball file?

Reply:
yeah, I think this can be improved in general. one catch is that we can't delete the tarball until it's pushed.